forward_grpc_test.go

```diff
@@ -0,0 +1,194 @@
+// spawning the goroutine.
+func unusedLocalTCPAddress(t testing.TB) string {
+	ln, err := net.Listen("tcp", "127.0.0.1:")
+	if err != nil {
+		t.Fatalf("Failed to bind to a test port: %v", err)
+	}
+	defer ln.Close()
+	return ln.Addr().String()
+}
+
+func testGRPCMetric(name string) string {
+	return grpcTestMetricPrefix + name
+}
+
+func forwardGRPCTestMetrics() []*samplers.UDPMetric {
+	return []*samplers.UDPMetric{
+		&samplers.UDPMetric{
+			MetricKey: samplers.MetricKey{
+				Name: testGRPCMetric("histogram"),
+				Type: histogramTypeName,
+			},
+			Value:      20.0,
+			Digest:     12345,
+			SampleRate: 1.0,
+			Scope:      samplers.MixedScope,
+		},
+		&samplers.UDPMetric{
+			MetricKey: samplers.MetricKey{
+				Name: testGRPCMetric("gauge"),
+				Type: gaugeTypeName,
+			},
+			Value:      1.0,
+			SampleRate: 1.0,
+			Scope:      samplers.GlobalOnly,
+		},
+		&samplers.UDPMetric{
+			MetricKey: samplers.MetricKey{
+				Name: testGRPCMetric("counter"),
+				Type: counterTypeName,
+			},
+			Value:      2.0,
+			SampleRate: 1.0,
+			Scope:      samplers.GlobalOnly,
+		},
+		&samplers.UDPMetric{
+			MetricKey: samplers.MetricKey{
+				Name: testGRPCMetric("timer"),
+				Type: timerTypeName,
+			},
+			Value:      100.0,
+			Digest:     12345,
+			SampleRate: 1.0,
+			Scope:      samplers.GlobalOnly,
+		},
+		&samplers.UDPMetric{
+			MetricKey: samplers.MetricKey{
+				Name: testGRPCMetric("set"),
+				Type: setTypeName,
+			},
+			Value:      "test",
+			SampleRate: 1.0,
+			Scope:      samplers.GlobalOnly,
+		},
+		// Only global metrics should be forwarded
+		&samplers.UDPMetric{
+			MetricKey: samplers.MetricKey{
+				Name: testGRPCMetric("counter.local"),
+				Type: counterTypeName,
+			},
+			Value:      100.0,
+			Digest:     12345,
+			SampleRate: 1.0,
+			Scope:      samplers.MixedScope,
+		},
+	}
+}
+
+func TestE2EForwardingGRPCMetrics(t *testing.T) {
+	ch := make(chan []samplers.InterMetric)
+	sink, _ := NewChannelMetricSink(ch)
+
+	ff := newForwardGRPCFixture(t, localConfig(), sink)
+	defer ff.stop()
+
+	input := forwardGRPCTestMetrics()
+	for _, metric := range input {
+		ff.IngestMetric(metric)
+	}
+	done := make(chan struct{})
+	go func() {
+		metrics := <-ch
+
+		expectedNames := []string{
+			testGRPCMetric("histogram.50percentile"),
+			testGRPCMetric("histogram.75percentile"),
+			testGRPCMetric("histogram.99percentile"),
+			testGRPCMetric("timer.50percentile"),
+			testGRPCMetric("timer.75percentile"),
+			testGRPCMetric("timer.99percentile"),
+			testGRPCMetric("counter"),
+			testGRPCMetric("gauge"),
+			testGRPCMetric("set"),
+		}
+
+		actualNames := make([]string, len(metrics))
+		for i, metric := range metrics {
+			actualNames[i] = metric.Name
+		}
+
+		assert.ElementsMatch(t, expectedNames, actualNames,
+			"The global Veneur didn't flush the right metrics")
+		close(done)
+	}()
+	ff.local.Flush(context.TODO())
+	ff.global.Flush(context.TODO())
```

Comment:
`context.Background()` for cases like this in tests, where we don't actually intend to use it for anything. (`context.TODO()` is intended to indicate that there is some future plan to plumb a proper context in).

Reply:
Sure, that makes a lot of sense.  I went through this diff and replaced usages of `context.TODO()` where they don't make sense. (b6f8298)